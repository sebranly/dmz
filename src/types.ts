// TODO: remove
export enum Template {
  Typescript = 'typescript',
  Javascript = 'javascript'
}

export type TemplateLanguage = {
  language: Template;
};
