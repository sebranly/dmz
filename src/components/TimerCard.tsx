import * as React from 'react';
import { DEAD_DROP_HOURLY_RATE, REGULAR_HOURLY_RATE } from '../constants/game';
import { NEW_RIBBON_DURATION_SEC } from '../constants/general';
import { Timer, TimeUnit } from '../types';
import { calculateRemainingSeconds, getEndTime } from '../utils';
import {
  convertSecondsToMoney,
  convertSecondsToTimeValue,
  convertTimerIndexToPlayerIndex,
  convertTimerIndexToPlayerTimerIndex
} from '../utils/convert';
import { displayWithTwoDigits, formatMoney, getPlayerColor, pluralize } from '../utils/display';
import classnames from 'classnames';

export interface TimerCardProps {
  className?: string;
  currentTimestamp: number;
  deadDropHourlyRate?: number;
  deleteTimer: () => void;
  regularHourlyRate?: number;
  timer: Timer;
}

const TimerCard: React.FC<TimerCardProps> = (props) => {
  const {
    className = '',
    currentTimestamp,
    deadDropHourlyRate = DEAD_DROP_HOURLY_RATE,
    deleteTimer,
    timer,
    regularHourlyRate = REGULAR_HOURLY_RATE
  } = props;
  const { timerIndex, timestampStart } = timer;

  const remainingSeconds = calculateRemainingSeconds(timer, currentTimestamp);
  const endTime = getEndTime(timer);

  const {
    [TimeUnit.Hour]: hours,
    [TimeUnit.Minute]: minutes,
    [TimeUnit.Second]: seconds
  } = convertSecondsToTimeValue(remainingSeconds);

  const isFixedHours = hours === 0;
  const isFixedMinutes = isFixedHours && minutes === 0;
  const isFixedSeconds = isFixedMinutes && seconds === 0;

  const playerIndex = convertTimerIndexToPlayerIndex(timerIndex);
  const playerTimerIndex = convertTimerIndexToPlayerTimerIndex(timerIndex);
  const color = getPlayerColor(playerIndex);
  const playerTitle = `Player ${playerIndex + 1}`;
  const timerTitle = `Timer ${playerTimerIndex + 1}`;
  const showRibbon = currentTimestamp - timestampStart <= NEW_RIBBON_DURATION_SEC;

  const classnamesPlayerColor = `text-${color}-500`;
  const classnamesPlayerTitle = classnames('timer-card-title', classnamesPlayerColor);
  const classnamesComponent = classnames(
    'timer-card-component',
    className,
    'border-solid',
    'border-2',
    'rounded-lg',
    `border-${color}-500`,
    'relative'
  );

  const classnamesMoney = classnames('money-value', classnamesPlayerColor);

  const items = [
    { value: hours, label: TimeUnit.Hour },
    { value: minutes, label: TimeUnit.Minute },
    { value: seconds, label: TimeUnit.Second }
  ].map((element: { value: number; label: string }) => {
    const { value, label: l } = element;
    const isFixed =
      (l === TimeUnit.Hour && isFixedHours) ||
      (l === TimeUnit.Minute && isFixedMinutes) ||
      (l === TimeUnit.Second && isFixedSeconds);

    const classnamesValue = classnames({ [classnamesPlayerColor]: isFixed });

    return (
      <li className="timer-card-element" key={l}>
        <span className={classnamesValue}>{displayWithTwoDigits(value)}</span>
        <div className="timer-card-unit">{pluralize(l, value)}</div>
      </li>
    );
  });

  return (
    <div className={classnamesComponent}>
      {showRibbon && <div className="ribbon-child">NEW</div>}
      <div className={classnamesPlayerTitle}>{playerTitle}</div>
      <div className="timer-card-title">{timerTitle}</div>
      <ul className="timer-card flex-container-timers">{items}</ul>
      <div className="timer-card-money">
        <div className="money-title">End Time:</div> <div className={classnamesMoney}>{endTime}</div>
      </div>
      <div className="timer-card-money">
        <div className="money-title">Exfiltration:</div>{' '}
        <div className={classnamesMoney}>
          ${formatMoney(convertSecondsToMoney(remainingSeconds, regularHourlyRate))}
        </div>
      </div>
      <div className="timer-card-money">
        <div className="money-title">Dead Drop:</div>{' '}
        <div className={classnamesMoney}>
          ${formatMoney(convertSecondsToMoney(remainingSeconds, deadDropHourlyRate))}
        </div>
      </div>
      <button
        className="mt-2.5 border-2 border-solid border-white text-base rounded-lg p-1 text-center bg-white text-black"
        onClick={() => deleteTimer()}
      >
        Delete this timer
      </button>
    </div>
  );
};

export { TimerCard };
