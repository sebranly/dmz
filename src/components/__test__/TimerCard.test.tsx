import * as React from 'react';
import { render } from '@testing-library/react';
import { TimerCard } from '../TimerCard';
import { Timer } from '../../types';

test('TimerCard', () => {
  const timer: Timer = { timerIndex: 0, durationSec: 62, timestampStart: 120 };
  const createComponent = () => (
    <TimerCard
      currentTimestamp={123}
      deadDropHourlyRate={30_000}
      deleteTimer={jest.fn()}
      regularHourlyRate={40_000}
      timer={timer}
    />
  );

  const { container } = render(createComponent());
  expect(container.childNodes).toMatchSnapshot();
});
