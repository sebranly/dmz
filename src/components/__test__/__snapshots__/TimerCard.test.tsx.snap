// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimerCard 1`] = `
NodeList [
  <div
    class="count-down  border-color-green ribbon-container"
  >
    <div
      class="ribbon-child"
    >
      NEW
    </div>
    <div
      class="count-down-title color-green"
    >
      Player 1
    </div>
    <div
      class="count-down-title"
    >
      Timer 1
    </div>
    <ul
      class="timer-card flex-container"
    >
      <li
        class="timer-card-element flex-child"
      >
        <span
          class="color-green timer-value"
        >
          00
        </span>
        <div
          class="timer-card-unit"
        >
          hours
        </div>
      </li>
      <li
        class="timer-card-element flex-child"
      >
        <span
          class="color-green timer-value"
        >
          00
        </span>
        <div
          class="timer-card-unit"
        >
          minutes
        </div>
      </li>
      <li
        class="timer-card-element flex-child"
      >
        <span
          class="timer-value"
        >
          59
        </span>
        <div
          class="timer-card-unit"
        >
          seconds
        </div>
      </li>
    </ul>
    <div
      class="timer-card-money"
    >
      <div
        class="money-title"
      >
        End Time:
      </div>
       
      <div
        class="money-value color-green"
      >
        01:03 AM
      </div>
    </div>
    <div
      class="timer-card-money"
    >
      <div
        class="money-title"
      >
        Exfil Money:
      </div>
       
      <div
        class="money-value color-green"
      >
        $
        700
      </div>
    </div>
    <div
      class="timer-card-money"
    >
      <div
        class="money-title"
      >
        Dead Drop:
      </div>
       
      <div
        class="money-value color-green"
      >
        $
        500
      </div>
    </div>
    <button
      class="remove-timer"
    >
      Delete this timer
    </button>
  </div>,
]
`;
